test/device/common.hpp

```diff
@@ -0,0 +1,227 @@
+#include <rocwmma/rocwmma.hpp>
+
+namespace rocwmma
+{
+    template <typename T>
+    __device__ inline float64_t toDouble(T const& val)
+    {
+        return static_cast<float64_t>(static_cast<float32_t>(val));
+    }
+
+    __device__ inline uint32_t rowMjr(uint32_t row, uint32_t col, uint32_t ld)
+    {
+        return row * ld + col;
+    }
+
+    __device__ inline uint32_t colMjr(uint32_t row, uint32_t col, uint32_t ld)
+    {
+        return col * ld + row;
+    }
+
+    __device__ inline float64_t maxDouble(float64_t a, float64_t b)
+    {
+        if (std::isinf(a) || std::isinf(b))
+        {
+            return std::numeric_limits<float64_t>::infinity();
+        }
+        // Check for NaN
+        else if (std::isnan(a) || std::isnan(b))
+        {
+            return std::numeric_limits<float64_t>::signaling_NaN();
+        }
+        return a > b ? a : b;
+    }
+
+    __global__ static void maxReduceKernel(float64_t* relativeError, uint32_t elements, uint32_t offset, uint32_t maxIdx)
+    {
+        float64_t* localRelativeError = relativeError + (offset * elements * blockIdx.x);
+
+        for (int i = elements >> 1; i > 0; i = i >> 1)
+        {
+            if (threadIdx.x < i && offset * (elements * blockIdx.x + threadIdx.x + i) < maxIdx)
+            {
+                localRelativeError[offset * threadIdx.x] = 
+                    maxDouble(localRelativeError[offset * threadIdx.x], localRelativeError[offset * (threadIdx.x + i)]);   
+            }
+            synchronize_workgroup();
+        }
+    }
+
+    // Comparitive kernel for batched matrix outputs as used in gemm tests
+    // Compares all values of two M x N matrices
+    template <typename TypeA, typename TypeB, typename LayoutA, typename LayoutB>
+    __global__ void compareEqualKernel(TypeA*   matrixA,
+                                       TypeB*   matrixB,
+                                       float64_t*  relativeError,
+                                       uint32_t m,
+                                       uint32_t n,
+                                       uint32_t lda,
+                                       uint32_t ldb)
+    {
+        uint32_t errorIdx = blockIdx.x * blockDim.x + threadIdx.x;
+        uint32_t colIdx = errorIdx % n;
+        uint32_t rowIdx = errorIdx / n;
+
+        uint32_t indexA = std::is_same<LayoutA, row_major>::value ? 
+            rowMjr(rowIdx, colIdx, lda) : colMjr(rowIdx, colIdx, lda);
+        uint32_t indexB = std::is_same<LayoutB, row_major>::value ? 
+            rowMjr(rowIdx, colIdx, ldb) : colMjr(rowIdx, colIdx, ldb);
+
+        if (rowIdx < m && colIdx < n)
+        {
+            TypeA valA = matrixA[indexA];
+            TypeB valB = matrixB[indexB];
+
+            // Determine relative error for each element of matrix A/B
+            auto numerator = fabs(toDouble(valA) - toDouble(valB));
+            auto divisor   = fabs(toDouble(valA)) + fabs(toDouble(valB)) + 1.0;
```

Comment:
Not sure if we should check NaN here, as we don't want to implicitly assume the maxReduce function will be called as a follow up. If it is signaling, I think there may be an exception thrown when you try to do the division.